wl-sim/main/WL_Flash.cpp

```diff
@@ -0,0 +1,292 @@
+    }
+
+    access_count = 0;
+
+    // copy data to dummy block
+    // data_addr = ...
+    // dummy_addr = ...
+    // erase_range(dummy_addr...
+    // for (copy count) read(...) write(...)
+    // done.. block moved
+    // write pos updates...
+
+    pos++;
+    if (pos >= MAX_POS) {
+        pos = 0;
+        // one loop more
+        move_count++;
+        if (move_count >= (MAX_POS - 1)) {
+            move_count = 0;
+            cycle_count++;
+        }
+    }
+    // write main state...
+
+    ESP_LOGD(TAG, "%s - access_count= 0x%08x, pos= 0x%08x, move_count= 0x%08x", __func__, (uint32_t) access_count, (uint32_t) pos, (uint32_t) move_count); 
+    return result;
+}
+
+esp_err_t erase_sector(size_t sector)
+{
+    esp_err_t result = ESP_OK;
+
+    updateWL();
+    size_t virt_addr = calcAddr(sector * SECTOR_SIZE);
+    size_t phy_sector = virt_addr / SECTOR_SIZE;
+
+    ESP_LOGV(TAG, "%s - virt_addr= 0x%08x, phy_sector= 0x%08x", __func__, (uint32_t) virt_addr, (uint32_t) phy_sector);
+
+    erase_count[phy_sector]++;
+
+    // reached maximum lifetime of a sector
+    // stop erasing and propagate to calculating normalized endurance (NE)
+    if (erase_count[phy_sector] >= SECTOR_ERASE_ENDURANCE) {
+        ESP_LOGI(TAG, "%s: sector %u reached %u", __func__, phy_sector, erase_count[phy_sector]);
+        return ESP_FAIL;
+    }
+
+    last_erase_sector = phy_sector;
+
+    return result;
+}
+
+// forward declaration
+void print_erase_counts(bool);
+
+esp_err_t erase_range(size_t start_address, size_t size)
+{
+    esp_err_t result = ESP_OK;
+
+
+    size_t erase_count = (size + SECTOR_SIZE - 1) / SECTOR_SIZE;
+    size_t start_sector = start_address / SECTOR_SIZE;
+
+    ESP_LOGD(TAG, "%s - start_address= 0x%08x, size= 0x%08x, erase_count= 0x%08x, start_sector= 0x%08x",
+            __func__, (uint32_t) start_address, (uint32_t) size, (uint32_t) erase_count, (uint32_t) start_sector); 
+
+    for (size_t i = 0; i < erase_count; i++) {
+        result = erase_sector(start_sector + i);
+        if (result != ESP_OK) {
+            // will propagate fail return code
+            break;
+        }
+    }
+
+    return result;
+}
+
+void print_erase_counts(bool verbose)
+{
+    uint64_t sum = 0;
+    uint32_t min = UINT32_MAX, max = 0, nonzeros = 0;
+
+    ESP_LOGI(TAG, "===== RESULTS =====");
+    for (size_t i = 0; i <= SECTOR_COUNT; i++) {
+        uint32_t count = erase_count[i];
+        if (count != 0) {
+            if (verbose) {
+                ESP_LOGI(TAG,"%ld: {%d}", i, count);
+            }
+            sum += count;
+            nonzeros++;
+
+            if (count < min) min = count;
+            if (count > max) max = count;
+        }
+    }
+    // normalized endurance [%]
+    //       Total Writes Before System Failure
+    // NE = ------------------------------------ x 100%
+    //               Wmax x Num Sectors
+    double NE = ((double)sum / (double)(SECTOR_ERASE_ENDURANCE * SECTOR_COUNT)) * 100;
```

Comment:
FYI: double data type is not supported at HW level for ESP32 chips, it's emulated software way => using doubles is expensive in terms of the CPU time. If possible, use float, which has HW implementation

Reply:
`wl-sim` is only to be run on PC, not on HW. even though it is an esp-idf project to be, it's supposed to be run only on PC `idf.py build monitor`. so double should be ok?